### PR TITLE
Add fuel efficiency tests

### PR DESCRIPTION
**Pull request by Coolthulhu**
Tests work as follows:
* Clean up map
* Spawn vehicle with 5% fuel (so that they don't take years to process)
* Turn its engine on and set cruise speed to safe speed
* Hack in player control so that it accelerates and doesn't skid
* Make it gain moves and idle
* Drag it back to starting point and measure distance
* Repeat last two points until engine dies
* Compare total measured distance to min and max

Results:
Fuel efficiency is BROKEN HARD!
The vehicle gets MORE efficient when it is offroading and when constantly accelerating and braking to 0.

On proper road, the car will use up the whole 5% of fuel tank to move 400 tiles (16 overmap tiles, 1.85 overmaps on full tank).
On dirt, the distance increases to 950.
When braking to full stop every 5 turns, it's 1250.
When braking to full stop every 5 turns on dirt, it's 1350.

Looks like the load on engine is calculated wrong (decreasing with traction) and the penalty for going fast greatly overpowers the bonus for optimal velocity. In fact, optimal velocity is horrible for efficiency.

### CODE REVIEW

**Comment by macrosblackd** [comment]
Can you add a test for idling? Fuel consumed per minute should be fine at 5% tank.

**Comment by Coolthulhu** [comment]
That would require a different test, for now I want a distance traveled test, because that is the most broken thing.

File: tests/vehicle_efficiency.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        reset_counter++;
+        if( reset_counter > reset_velocity_turn ) {
+            veh.velocity = 0;
+            veh.last_turn = 0;
+            veh.of_turn_carry = 0;
+            reset_counter = 0;
+        }
+    }
+
```
Comment (kevingranade) [inline tests/vehicle_efficiency.cpp]: Adding something like
INFO( "vehicle: " << veh_id.str() << ". Terrain: " << terrain << "." );
here helps with debugging if it does fail.

**Comment by kevingranade** [comment]
If we keep having more of these we might eventually be pretty resistant to regressions, awesome.

Would we call efficiency good or at least better if these tests passed?  Or are these just demonstrating that the relationship between these scenarios is bad and that should be fixed?
If it's the latter I'd recommend doing something like:
Establish distance travelled on road without halting.
Verify that distance travelled on dirt without halting is proportional (half? not sure)
Verify that distance travelled on road with braking/halting is proportional.
Verify that distance travelled on dirt with braking/halting is proportional.

With this method we can fix the problem and commit the test as a regression checker without worrying about what the absolute value should be.

**Comment by Coolthulhu** [comment]
> Would we call efficiency good or at least better if these tests passed?

That has to be figured out. The tests like this are very helpful with figuring that out.
Before I wrote those, I tested fuel efficiency by starting a mostly empty world and driving through it in game. This takes a lot of time and requires supervision due to the world not being perfectly empty.

> Verify that distance travelled on dirt without halting is proportional (half? not sure)

I would avoid specifying this too accurately. The underlying math is "iterative" and thus not that easy to track.
Still, it should probably be within 50%-75%.

For now the values are just placeholders - they could be changed to just verify that worse terrain and driving habits result in worse efficiency.

**Comment by kevingranade** [comment]
Verify that distance travelled on dirt without halting is proportional
(half? not sure)

I would avoid specifying this too accurately. The underlying math is
"iterative" and thus not that easy to track.

I was being lazy with my phrasing, the acceptance criteria would I
something like "continuous driving on dirt uses 40% to %60 more fuel than
on road. "
I'm just looking for a regression test we'd be OK to merge along with a
fix, it doesn't matter if it's the final numbers, but it gets us to a place
where we can discuss actual numbers instead of handwave-y approximations.

**Comment by kevingranade** [comment]
I backported this to pre-overhaul code, and the behavior is very strange.
It's bailing out of a test way before the fuel is exausted, I suspect the cruise calculation may be off and its wrecking the engine.  No idea if this is happening with recent master.

**Comment by kevingranade** [comment]
confirmed, set cruise speed to safe speed insted of 223x safe speed and it works much more as expected.  I'm looking into migrating the now-working tests to master to look into tweaking values for efficiency.

**Comment by keyspace** [comment]
Perhaps can be closed, since superceded by #20007?..

**Pull request closed without merging**